win32/getopt.h

```diff
@@ -0,0 +1,902 @@
+     */
+    for( --optmark; optmark >= optbase; --optmark )
+      arglist[optmark + optspan] = arglist[optmark];
+
+    /* restore the temporarily saved option arguments to
+     * their permuted location.
+     */
+    for( index = 0; index < optspan; ++index )
+      arglist[optbase + index] = this_arg[index];
+
+    /* adjust `optbase', to account for the relocated option.
+     */
+    optbase += optspan;
+	
+	free(this_arg);
+  }
+
+  else
+    /* no permutation occurred ...
+     * simply adjust `optbase' for all options parsed so far.
+     */
+    optbase = argind + 1;
+
+  /* enter main parsing loop ...
+   */
+  while( argc > ++argind )
+  {
+    /* inspect each argument in turn, identifying possible options ...
+     */
+    if( is_switchar( *(nextchar = argv[optmark = argind]) ) && *++nextchar )
+    {
+      /* we've found a candidate option argument ... */
+
+      if( is_switchar( *nextchar ) )
+      {
+	/* it's a double hyphen argument ... */
+
+	const CHAR *refchar = nextchar;
+	if( *++refchar )
+	{
+	  /* and it looks like a long format option ...
+	   * `getopt_long' mode must be active to accept it as such,
+	   * `getopt_long_only' also qualifies, but we must downgrade
+	   * it to force explicit handling as a long format option.
+	   */
+	  if( mode >= getopt_mode_long )
+	  {
+	    nextchar = refchar;
+	    mode = getopt_mode_long;
+	  }
+	}
+	else
+	{
+	  /* this is an explicit `--' end of options marker, so wrap up now!
+	   */
+	  if( optmark > optbase )
+	  {
+	    /* permuting the argument list as necessary ...
+	     * (note use of `this_arg' and `arglist', as above).
+	     */
+	    CHAR *this_arg = argv[optmark];
+	    CHAR **arglist = (CHAR **)(argv);
+
+	    /* move all preceding non-option arguments to the right ...
+	     */
+	    do arglist[optmark] = arglist[optmark - 1];
+	       while( optmark-- > optbase );
+
+	    /* reinstate the `--' marker, in its permuted location.
+	     */
+	    arglist[optbase] = this_arg;
+	  }
+	  /* ... before finally bumping `optbase' past the `--' marker,
+	   * and returning the `all done' completion indicator.
+	   */
+	  optind = ++optbase;
+	  return getopt_all_done;
+	}
+      }
+      else if( mode < getopt_mode_long_only )
+      {
+	/* it's not an explicit long option, and `getopt_long_only' isn't active,
+	 * so we must explicitly try to match it as a short option.
+	 */
+	mode = getopt_mode_standard;
+      }
+
+      if( mode >= getopt_mode_long )
+      {
+	/* the current argument is a long form option, (either explicitly,
+	 * introduced by a double hyphen, or implicitly because we were called
+	 * by `getopt_long_only'); this is where we parse it.
+	 */
+	int lookup;
+	int matched = -1;
+
+	/* we need to fetch the `extra' function arguments, which are
+	 * specified for the `getopt_long' APIs.
+	 */
+	va_list refptr;
+	va_start( refptr, optstring );
+	struct option *longopts = va_arg( refptr, struct option * );
+	int *optindex = va_arg( refptr, int * );
```

Comment:
Declare the variables `longopts` and `optindex` above the `va_start` call, to fix compilation with MSVC 2012 and earlier.